### ISSUE TITLE
ATLauncher packs

### ISSUE THREAD

**Issue by thesymbol** [hexparrot/mineos-node#86]
If you got time over to add more modpacks, would be nice to have ATLauncher modpacks included aswell.

They have got an API as far as i could see here:
http://wiki.atlauncher.com/api:packs

**Comment by hexparrot** [comment]
I definitely have the time to add more modpacks, but actually I have no idea what I'm looking at with ATLauncher.  Not looking for code, but can you provide me a bit to get me started in what URLs for example I'd be able to use to get a sample pack up?

**Comment by thesymbol** [comment]
as far as i have looked into it (just for a couple of minutes).
the urls for the website are (and API):
* https://api.atlauncher.com/v1/ (API, it supports RESTful functions and returns in JSON Objects)
* https://atlauncher.com/ (website)
* https://wiki.atlauncher.com/ (documentation)

there does also seem to be a nodejs module for working with the ATLauncher API here (not looked into it):
https://www.npmjs.com/package/atlauncher-api

to select a specific pack:
https://api.atlauncher.com/v1/pack/[pack name]
for example getting Sky Factory:
https://api.atlauncher.com/v1/pack/SkyFactory
it lists the different versions and url for each version to download.

to get the list of all packs:
https://api.atlauncher.com/v1/packs/full/all/

ATLauncher has a limit on the amount of API calls done to their website. 300 calls per hour for the public API.
Pack developer has 3000 API calls
and Partner has 600 API calls

NOTE: i edited the example output to keep it a bit smaller than the actual api return value cause otherwise it would take up this whole page.
Example output for packs/full/all:
```
{
    "error": false,
    "code": 200,
    "message": null,
    "data": [
        {
            "id": 1,
            "name": "MoonQuest",
            "safeName": "MoonQuest",
            "type": "public",
            "versions": [
                {
                    "version": "0.17.0",
                    "minecraft": "1.6.4",
                    "published": 1394492435,
                    "__LINK": "https://api.atlauncher.com/v1/pack/MoonQuest/0.17.0"
                },
                {
                    "version": "0.16.0",
                    "minecraft": "1.6.4",
                    "published": 1393246055,
                    "__LINK": "https://api.atlauncher.com/v1/pack/MoonQuest/0.16.0"
                }
            ],
            "description": "Welcome to this Yogscast Inspired Unofficial Minecraft Modpack! If you haven't seen their new Minecraft series yet this modpack has all those mods and now you can go ahead and play along too! You should really go watch it!",
            "supportURL": "http://forums.atlauncher.com/forum-37.html",
            "websiteURL": "http://www.atlauncher.com/packs/MoonQuest/"
        },
        {
            "id": 2,
            "name": "Magicality: The Next Dimension",
            "safeName": "MagicalityTheNextDimension",
            "type": "public",
            "versions": [
                {
                    "version": "ALPHA-2.0",
                    "minecraft": "1.7.10",
                    "published": 1417883998,
                    "__LINK": "https://api.atlauncher.com/v1/pack/MagicalityTheNextDimension/ALPHA-2.0"
                },
                {
                    "version": "ALPHA-1.0",
                    "minecraft": "1.7.10",
                    "published": 1417623248,
                    "__LINK": "https://api.atlauncher.com/v1/pack/MagicalityTheNextDimension/ALPHA-1.0"
                }
            ],
            "description": "Welcome to 1.7.10 and welcome Magicality into the HQM family! Magicality is a heavily story driven (thanks to Lore Expansion) and heavily edited Minecraft experience. It contains many magical mods (no pesky Tech mods in sight!) follow your own story and lead the fight against those block stealing foe's. Please note that it is currently in Alpha things may change. There is NO overworld ore generation, no use looking for it! ",
            "supportURL": "http://forums.atlauncher.com/forum-29.html",
            "websiteURL": "http://forums.atlauncher.com/forum-29.html"
        },
    ]
}
```

Example output for specific pack:
```
{
    "error": false,
    "code": 200,
    "message": null,
    "data": {
        "id": 1,
        "name": "MoonQuest",
        "safeName": "MoonQuest",
        "type": "public",
        "versions": [
            {
                "version": "0.17.0",
                "minecraft": "1.6.4",
                "published": 1394492435,
                "__LINK": "https://api.atlauncher.com/v1/pack/MoonQuest/0.17.0"
            },
            {
                "version": "0.16.0",
                "minecraft": "1.6.4",
                "published": 1393246055,
                "__LINK": "https://api.atlauncher.com/v1/pack/MoonQuest/0.16.0"
            }
        ],
        "description": "Welcome to this Yogscast Inspired Unofficial Minecraft Modpack! If you haven't seen their new Minecraft series yet this modpack has all those mods and now you can go ahead and play along too! You should really go watch it!",
        "supportURL": "http://forums.atlauncher.com/forum-37.html",
        "websiteURL": "http://www.atlauncher.com/packs/MoonQuest/"
    }
}
```

**Comment by thesymbol** [comment]
Quick update
here is an example for a version request
https://api.atlauncher.com/v1/pack/SkyFactory/[version number]
example
https://api.atlauncher.com/v1/pack/SkyFactory/2.4

the field "serverZipURL" will be null if no pack was found.

example output:
```
{
    "error": false,
    "code": 200,
    "message": null,
    "data": {
        "version": "2.4",
        "minecraftVersion": "1.7.10",
        "recommended": true,
        "published": 1431306473,
        "serverZipURL": "http://fs.atlcdn.net/downloads/servers/public/SkyFactory/2.4/SkyFactory - 2.4.zip?hash=1iV7ceUhDjF6uhmT8hJLBA&expires=1439145440",
        "changelog": "=====================================\r\nAlways back up your worlds before updating!\r\n=====================================\r\n\r\nWelcome to Sky Factory 2! A lot has changed since version 1. For tutorials see my YouTube channel:\r\nhttp://YouTube.com/BaconDonutTV\r\n\r\nOr watch me play this pack and others live over at:\r\n http://twitch.tv/bacon_donut\r\n\r\n=====================================\r\n\r\nSky Factory 2.4 - The Magical Uprising\r\n---------------------\r\n\r\nRemoved Mods:\r\n\r\nCrafting Manager\r\n\r\n\r\nMods Added:\r\n\r\nMineTweaker \r\nMod Tweaker 2\r\nAE2 Stuff\r\nAutomagy\r\nBotania\r\nCosmetic Armor\r\nDecoCraft\r\nForbidden Magic\r\nMFR Magical Crops\r\nRanchable Fluid Cows\r\nRedLogic\r\nRFTools\r\nThaumcraft\r\nThaumcraft NEI Plugin\r\nThaumCraft Node Tracker\r\nThaumic Energistics\r\nThermal Casting\r\nTranslocators\r\nWitchery\r\n\r\nMods Updated:\r\n\r\nApplied Energistics\r\nBiblioCraft\r\nBig Reactors\r\nCarpenter's Blocks \r\nChisel 2 \r\nCoFHCore\r\nDraconic Evolution\r\nEnder IO \r\nEnderTech \r\nEx-Astris \r\nEx-Nihilo\r\nExtraCells 2\r\nExtra Utilities \r\nIguana Tinker Tweaks \r\nImmibis Core \r\nInfiniteInvo\r\nMagical Crops  \r\nMineFactory Reloaded\r\nNEI Intergration\r\nOpen Blocks \r\nOpenModsLib \r\nOpenPeripheralAddons \r\nOpenPeripheralCore \r\nOpenPeripheralIntergration \r\nRedstone Arsenal\r\nSolarFlux\r\nThermal Dynamics \r\nThermal Expansion\r\nThermal Foundation \r\nttCore\r\nVeinMiner \r\n\r\nConfig Changes:\r\n\r\nFixed Configs that were changed with the last update.\r\nDisabled Tinkers Copper and Tin ore generation.\r\nAdded recipe for Iron Golem Spawn Egg to use for a stable sigil ritual."
    }
}
```

**Comment by hexparrot** [comment]
It seems like there's a bit of technical issue when it comes to adding ATLauncher stuff to the webui.

Most notably, I can't seem to consistently get a list of verifiably-downloadable packs--or at least in a way that wouldn't be horribly over-reaching and over-processing.  I'll provide an example:

https://api.atlauncher.com/v1/packs/full/public/

This is the URL I am using to get a full list of publicly-available packs.  It returns once nicely formatted and is a manage-able 71KB. Then I'll take the very first result:

```
{
    "error": false,
    "code": 200,
    "message": null,
    "data": [
        {
            "id": 1,
            "name": "MoonQuest",
            "safeName": "MoonQuest",
            "type": "public",
            "versions": [
                {
                    "version": "0.17.0",
                    "minecraft": "1.6.4",
                    "published": 1394492435,
                    "__LINK": "https://api.atlauncher.com/v1/pack/MoonQuest/0.17.0"
                },
```

Using the __LINK, I can download it and get:

```
{
    "error": false,
    "code": 200,
    "message": null,
    "data": {
        "version": "0.17.0",
        "minecraftVersion": "1.6.4",
        "recommended": true,
        "published": 1394492435,
        "serverZipURL": null,
        "changelog": "...."
    }
}
```

But then, as you state, serverZipURL is empty!  So the problem that I have is: how am I to know that a server pack cannot actually be downloaded *before* making a second, separate call to __LINK?

If the Webui shows all the packs (and all the versions) and then the user clicks "download"...the backend has no problem making the second API call...but ultimately, no download will occur.  If no download will occur (as is the case for a vast majority of the links of packs), then it shouldn't be listed.  But if the only way for me to know it's not available, is to make the API call...then the webui will also have to make dozens/hundreds of API calls *before* an admin even knows he wants one version of all those available.

TLDR; is there a way to verify the availability of a pack without an API call?

**Comment by flareofghast** [comment]
An idea is that you could maintain a list of available packs. Have an update button to run through and check them all. I know it's a pain, just giving ideas.

**Comment by thesymbol** [comment]
Have to agree with flareofghast is to keep a cached version in .json format and let the user update the .json manually by hitting an update button.

Will look further into it but doubt you can get a full list with versions & download links for the zip files.

**Comment by hexparrot** [comment]
Actually, it's not technically *difficult* to do this, it is technically *wasteful*/slow/undesirable.  As the OP wrote in his first post:

> ATLauncher has a limit on the amount of API calls done to their website. 300 calls per hour for the public API.

Each time you start the webui up (or manually update profiles), MineOS downloads all the XML/JSON/etc files for each profile group (vanilla, ftb, pocketmine).  This is a total of about 3 or 4 requests, which are then parsed and sent to the client; this is pretty lightweight.

One plus is that these downloaded files contain consistent, dependable URLs for the corresponding server packs.  On the other hand, the master list is only 1 additional request, but it lists more than 300 packs from that single file.

So the biggest concerns are:

1) of the hundreds of profiles the master list makes available, not all of them actually have a URL
2) one can't know whether it has a URL until you make a second, individual request for that pack
3) checking all packs would exceed the 300 per hour
4) even if it were *only* 300 packs total (or less than the limit), maybe 99.9% of MineOS webui start ups won't involve downloading one of these packs, making a LOT of unused (thus unnecessary) web requests.

So yes, it's definitely possible to have a full list--(1 request)--and allow the user to hit 'update' to get any one individual.  The only concern here is that I have found that 90% or more of the available packs don't have a URL.  So 90% of the users clicks might end up with a "this pack isn't available" error, in the end anyway.

**Comment by thesymbol** [comment]
Well since ATLaunchers API isnt built to be used for listing download links.
It would be a great idea to allow downloading of profiles from custom URLs, like you enter a name for the profile, then specify if its a .zip or a .jar and then input the URL for the file to download or uploading from your local computer with the same settings except for the url witch is replaced by a input field for uploading a file.

**Comment by hexparrot** [comment]
Except that brings us back into Profiles (from Python MineOS) territory.  Though it may not be readily-apparent, I personally have received so many emails asking how to get profiles to work with custom URLs that it necessitated redesigning them altogether.

Allowing custom URLs then requires that all that information gets saved somewhere (previously profiles.config) and there's parsing/user error that needs to be anticipated, as well as logic to handle how it gets unzipped, and whether a parent, containing directory exists and needs to be relocated...it gets messy.

As is on mineos-node, if it doesn't exist on a profile, you have a very simple task of uploading and unzipping the file into /var/games/minecraft/servers/myservername.  That's it.  

Don't get me wrong.  I would prefer that command line, filezilla, etc. never have to be used to get things done with MineOS, but understand too that when some infrastructure makes it too difficult (e.g., having an API for downloads and having expiration parameters in the URL), that means it just offloads all the work for me to do.  And while I love coding for this project, ATLauncher has had so little mention in my time with Python-based MineOS that I'm not sure I can justify the LOTs of code it would require and the inevitable design changes and complexity increases that would accompany it.

Put another way, I don't think ATLauncher is popular enough for me to change the *whole design of how profiles work*, since the alternative is very simple with the new handling of "profiles."

**Comment by thesymbol** [comment]
I get what your saying, i can understand that for you to redesign the entier profile system around ATLauncher would be a mess and would only make things worse.

As for custom uploads/profiles it would be a nice to have feature but i get what your saying it is an aweful amount of work to make sure that the user do not input the wrong url and that the zip doesnt need to be handled in a different way than others.

The main reason behind the custom profiles would be to prefferebly not have to open up FileZilla or ssh into the server since that is a pain to do alot of the times.

**Comment by Zone22** [comment]
@hexparrot I was there near the beginning of ATLauncher back in the day when it was just used for their server. That being said I could reach out to them. In doing so I would need a concise summary of the problem that I could pass along.

**Comment by hexparrot** [comment]
I doubt they are very interested in changing their workflow to accommodate MineOS automation.  The API itself is designed specifically to funnel users to their website, using their API and interface.  There's not a lot of incentive for them to change any of that, but lots to lose, e.g., ad revenue, centralized internet traffic, branding...
